swarm/pss/prox_test.go

```diff
@@ -255,93 +243,153 @@ func testProxNetwork(t *testing.T) {
 		// context deadline exceeded
 		// however, it might just mean that not all possible messages are received
 		// now we must check if all required messages are received
-		cnt := tstdata.getMsgCount()
-		log.Debug("TestProxNetwork finished", "rcv", cnt)
-		if cnt < tstdata.requiredMessages {
+		log.Debug("TestProxNetwork finished", "rcv", tstdata.getMsgCount())
+		if !isSuccess(tstdata) {
 			t.Fatal(result.Error)
 		}
 	}
-	t.Logf("completed %d", result.Duration)
 }
 
-func (tstdata *testData) sendAllMsgs() {
+func isSuccess(d *testData) bool {
+	cnt := d.getMsgCount()
+	if cnt < d.requiredMessages {
+		if d.isExpectationsChanged() {
+			// network configuration has changed since beginning of
+			// the test, which invalidates our original expectations
+			log.Warn("expectations changed")
+			return true
+		} else {
+			return false
+		}
+	} else {
+		return true
+	}
+}
+
+func (d *testData) isExpectationsChanged() bool {
+	expected := make(map[enode.ID][]uint64)
+	for i := 0; i < len(d.msgs); i++ {
+		var targets []enode.ID
+		var closestPO int
+		for _, nod := range d.sim.Net.GetNodes() {
+			po, _ := pof(d.msgs[i], d.nodeAddrs[nod.ID()], 0)
+			if po > closestPO {
+				closestPO = po
+				targets = nil
+				targets = append(targets, nod.ID())
+			} else if po == closestPO {
+				targets = append(targets, nod.ID())
+			}
+		}
+		for _, id := range targets {
+			expected[id] = append(expected[id], uint64(i))
+		}
+	}
+	changed := !reflect.DeepEqual(d.requiredMessages, expected)
+	if changed {
+		log.Warn("expectations changed")
+	}
+	return changed
+}
+
+func (tstdata *testData) sendAllMsgs(nodes map[int]*rpc.Client) {
 	for i, msg := range tstdata.msgs {
 		log.Debug("sending msg", "idx", i, "from", tstdata.senders[i])
-		nodeClient, err := tstdata.sim.Net.GetNode(tstdata.senders[i]).Client()
-		if err != nil {
-			tstdata.errC <- err
-		}
+		nodeClient := nodes[i]
 		var uvarByte [8]byte
 		binary.PutUvarint(uvarByte[:], uint64(i))
 		nodeClient.Call(nil, "pss_sendRaw", hexutil.Encode(msg), hexutil.Encode(topic[:]), hexutil.Encode(uvarByte[:]))
 	}
-	log.Debug("all messages sent")
 }
 
 // testRoutine is the main test function, called by Simulation.Run()
-func testRoutine(tstdata *testData, ctx context.Context) error {
-	go handlerChannelListener(tstdata, ctx)
-	go tstdata.sendAllMsgs()
+func testRoutine(d *testData, ctx context.Context) error {
+	go handlerChannelListener(d)
```

Comment:
I don't find that intuitive that the passed testData has a channel which we use to terminate this routine.
Could we have `handlerChannelListener()` maybe on `testData`? I wonder if that would help.

Also note having `<-ctx.Done` in a goroutine does not guarantee clean termination. Just termination.  Better than nothing as we don't leak resources. If you want clean termination you need a "callback" channel what you can wait on.

Maybe - without understanding the whole context - I would define a `waitGroup` on `testData`; move all go routine starts into a method on `testData` and add a `testData.Stop()` (?naming?) method that would cancel context and could call `wg.Wait()`.